File: java/purchase-microservice/src/main/java/purchases/purchasemicroservice/service/PurchaseService.java
package purchases.purchasemicroservice.service;

import org.springframework.stereotype.Service;
import org.springframework.web.client.RestTemplate;
import purchases.purchasemicroservice.model.PurchaseEvent;
import purchases.purchasemicroservice.repository.PurchaseRepository;
import org.springframework.beans.factory.annotation.Value;

import java.util.List;

@Service
public class PurchaseService {

    private final PurchaseRepository purchaseRepository;
    private final RestTemplate restTemplate;
    private final String apiUrl;

    public PurchaseService(PurchaseRepository purchaseRepository, RestTemplate restTemplate, @Value("${API_BASE_URL}") String apiUrl) {
        this.purchaseRepository = purchaseRepository;
        this.restTemplate = restTemplate;
        this.apiUrl = apiUrl;
    }

    public PurchaseEvent save(PurchaseEvent purchaseEvent) {
        return purchaseRepository.save(purchaseEvent);
    }

    public List<PurchaseEvent> getAll() {
        return purchaseRepository.findAll();
    }

    public PurchaseEvent get(String id) {
        return purchaseRepository.findById(id).orElseThrow(() -> new RuntimeException("PurchaseEvent not found"));
    }

    public void delete() {
        purchaseRepository.deleteAll();
    }

    public Float getTaxValue(String pole) {
        Float taxValue = restTemplate.getForObject(apiUrl + "/tax/" + pole, Float.class);

    }
}

File: fastapi/purchase-microservice/services/purchase_service.py
from typing import List

import requests

from fastapi import FastAPI

from ..model.purchase_event import PurchaseEventModel
from ..repository.purchase_repository import PurchaseRepository


class PurchaseService:
    def __init__(self, app: FastAPI):
        self.purchase_repository: PurchaseRepository = PurchaseRepository(app)
        self.tax_microservice_url: str = app.state.config['TAX_MICROSERVICE_URL']

    async def _save(self, purchase_event: PurchaseEventModel) -> PurchaseEventModel:
        return self.purchase_repository.save(purchase_event)

    async def _get_tax_value(self, region: str) -> float:
        return requests.get(url=self.tax_microservice_url + '/tax/' + region).json()

    async def _calculate_full_price(self, house_price: float, tax: float) -> float:
        return house_price + (house_price * tax)/100

    async def register(self, purchase_event: PurchaseEventModel) -> PurchaseEventModel:
        tax_value: float = await self._get_tax_value(purchase_event['region'])
        fullPrice: float = await self._calculate_full_price(float(purchase_event['price']), tax_value)
        purchase_event['price'] = fullPrice
        return await self._save(purchase_event)

    async def get_all(self) -> List[PurchaseEventModel]:
        return await self.purchase_repository.find_all()

    async def delete_all(self) -> None:
        await self.purchase_repository.delete_all()
